### PR TITLE
Fix #4835: Treat excluded classes for Tagger as path roots.

### LINKED ISSUE #4833: High memory consumption during linking when using `ModuleSplitStyle.SmallModulesFor(...)`
This is based on a [discord discussion](https://discord.com/channels/632150470000902164/635668814956068864/1084860483305222254).

We have a webapp using  ModuleSplitStyle.SmallModulesFor(...). 
Scala 3.2.2, Scalajs 1.13

`fullLinkJS` (and `fastLinkJS`) produces a 7GB HashMap full of 50 million `linker.frontend.modulesplitter.Tagger$DynamicPaths`.
Which makes the compilation painfully slow.

No issues with the default split style `ModuleSplitStyle.FewesModules`, and also no issue with `ModuleSplitStyle.SmallestModules`

A tree that has the issue (not minimized at all) is available here: 
https://github.com/rmgk/reform-scalajs-linker-memory-issue
`sbt fullLinkJS` is enough to trigger the issue. I tested the split style by changing line 36 `build.sbt`.

@gzm0 tagging you as you requested this issue as a reminder (and I don’t think I can assign people) :-)

### LINKED ISSUE #4835: SmallModulesFor splitting style can cause circular dependencies between emitted modules
Minimal example:

```scala
import scala.scalajs.js.annotation._

package helloworld {
  object HelloWorld {
    def main(args: Array[String]): Unit = {
      println(A.baz())
    }

    @JSExportTopLevel("yolo", "mod")
    def exp() = A.baz()
  }

  object A {
    @noinline
    def baz() = app.A.foo() + B.bar()
  }

  object B {
    @noinline
    def bar() = 1
  }
}

package app {
  object A {
    @noinline
    def foo() = helloworld.B.bar() + 1
  }
}
```

Observation:

```
[error] (helloworld2_12 / Compile / fastLinkJS) java.lang.IllegalArgumentException: requirement failed: Must have exactly one root module
```

Which is this assertion here: https://github.com/scala-js/scala-js/blob/44014be9ef5904f3b19052ac5c9861b1020567c4/linker/shared/src/main/scala/org/scalajs/linker/standard/ModuleSet.scala#L47-L48

Disabling it, we can observe that we generate the following modules:

```
app.A$.js.map
internal-30d24927543dc20e4b7b5599dbff65130a4a6a4a.js
main.js
mod.js
```

And the two internal modules depend on each other.

This is essentially the degenerate case described here: https://github.com/scala-js/scala-js/issues/4833#issuecomment-1475252934

The reason this happens is that the SmallModulesFor analyzer abuses dynamic edges to avoid this scenario. But at the same time, dynamic edges analyze possible execution flows to avoid oversplitting.

### PR DESCRIPTION
**Pull request by sjrd**
Previously, we more or less abused the concept of dynamic dependencies to handle them. This was based on the observation that we "must" split modules at those boundaries. See https://github.com/scala-js/scala-js/issues/4327#issuecomment-1059984564

However, as #4835 highlights, that observation ignored one aspect of dynamic dependencies: that when A--dyn-->B, A--static-->C and B--static-->C, by the time we load `B`, we know that `C` has already been loaded, and it is therefore OK to put A and C in the same module but not B.

For the excluded classes for `SmallModulesFor`, we cannot make that assumption. On the contrary, relying on it causes a circular dependency between the A-C module and the B module.

Instead, we now handle excluded classes as additional path *roots* of the analysis, similarly to entry points. That still encourages them to be put in separate modules, but it can get overruled when regular static dependencies demand that they stick together with other classes to prevent cycles.

This commit also fixes #4833.

### CODE REVIEW

**Comment by sjrd** [comment]
@gzm0 Does the commit message make any sense? Did I understand correctly what is happening and why your changes fix the issue?

Also I verified that it fixes #4833 on the provided reproduction, but I have no idea why...

**Comment by sjrd** [comment]
It does make the module analyzer 2x slower in the non-degenerate case of the test suite, though.

Before this commit (incremental run after changing one test file):
```
sbt:Scala.js> last testSuite2_12/Test/fastLinkJS
[info] Fast optimizing /localhome/doeraene/projects/scalajs/test-suite/js/.2.12/target/scala-2.12/scalajs-test-suite-test-fastopt
[debug] Linker: Compute reachability: 357161 us
[debug] Linker: Assemble LinkedClasses: 98817 us
[debug] Linker: Check IR: 254917 us
[debug] Linker: 736357 us
[debug] Optimizer: Batch mode: false
[debug] Optimizer: Incremental part: 114043 us
[debug] Optimizer: Optimizing 72 methods.
[debug] Optimizer: Optimizer part: 10280 us
[debug] Optimizer: 143166 us
[debug] Refiner: Compute reachability: 200735 us
[debug] Refiner: Assemble LinkedClasses: 26855 us
[debug] Refiner: 247074 us
[debug] Module Splitter: Calculate Dependency Info: 2343 us
[debug] Module Splitter: Analyze Modules: 3028779 us
[debug] Module Splitter: Assemble Modules: 13523 us
[debug] Module Splitter: 3044924 us
[debug] Emitter: Generate Classes: 82543 us
[debug] Emitter: Write trees: 24235 us
[debug] Emitter: Class tree cache stats: reused: 3552 -- invalidated: 3
[debug] Emitter: Method tree cache stats: reused: 32321 -- invalidated: 55
[debug] Emitter: 151501 us
[debug] BasicBackend: Write result: 62121 us
[debug] BasicBackend: total top-level trees: 87186; re-computed: 132
[debug] BasicBackend: total modules: 2389; re-written: 3
```
After this commit (same changes):
```
sbt:Scala.js> last testSuite2_12/Test/fastLinkJS
[info] Fast optimizing /localhome/doeraene/projects/scalajs/test-suite/js/.2.12/target/scala-2.12/scalajs-test-suite-test-fastopt
[debug] Linker: Compute reachability: 217953 us
[debug] Linker: Assemble LinkedClasses: 60553 us
[debug] Linker: Check IR: 123716 us
[debug] Linker: 423710 us
[debug] Optimizer: Batch mode: false
[debug] Optimizer: Incremental part: 57192 us
[debug] Optimizer: Optimizing 72 methods.
[debug] Optimizer: Optimizer part: 9110 us
[debug] Optimizer: 95796 us
[debug] Refiner: Compute reachability: 322820 us
[debug] Refiner: Assemble LinkedClasses: 34343 us
[debug] Refiner: 372757 us
[debug] Module Splitter: Calculate Dependency Info: 2331 us
[debug] Module Splitter: Analyze Modules: 6070918 us
[debug] Module Splitter: Assemble Modules: 11139 us
[debug] Module Splitter: 6084598 us
[debug] Emitter: Generate Classes: 44029 us
[debug] Emitter: Write trees: 20994 us
[debug] Emitter: Class tree cache stats: reused: 3552 -- invalidated: 3
[debug] Emitter: Method tree cache stats: reused: 32321 -- invalidated: 55
[debug] Emitter: 97159 us
[debug] BasicBackend: Write result: 37988 us
[debug] BasicBackend: total top-level trees: 90427; re-computed: 132
[debug] BasicBackend: total modules: 2445; re-written: 3
```
Looks like I know what I will point flame graphs at, next time I get into an optimization frenzy. :sweat_smile:

**Comment by gzm0** [comment]
> Also I verified that it fixes https://github.com/scala-js/scala-js/issues/4833 on the provided reproduction, but I have no idea why...

The reason is that as soon as we actually trigger the fake dynamic edge, we "enter" the excluded classes. As soon as that happens, we'll create fake dynamic edges for the entire dependency graph of the excluded classes (you'll note that we'll always create a fake dynamic edge when we go to an excluded class, irrespective of where we're coming from).

File: linker/shared/src/test/scala/org/scalajs/linker/SmallModulesForSplittingTest.scala [hunk 1]
```diff
@@ -92,4 +94,87 @@ class SmallModulesForSplittingTest {
...
+      EntryPointsClass,
+      superClass = Some(ObjectClass),
+      methods = List(
+        trivialCtor(EntryPointsClass),
+        mainMethodDef(consoleLog(call("lib.A", "baz")))
+      ),
+      topLevelExportDefs = List(
+        TopLevelMethodExportDef("mod",
+            JSMethodDef(SMF, str("exp"), Nil, None, call("lib.A", "baz"))(EOH, UNV))
+      )
```
Comment (gzm0) [inline linker/shared/src/test/scala/org/scalajs/linker/SmallModulesForSplittingTest.scala:133]: Consider just two top-level method export defs for simplicity / symmetry. FTR: The reason we need two entry points is that `lib.A` doesn't go into the public module (and `lib.B` into a different one due to public module import avoidance).

**Comment by gzm0** [comment]
> Does the commit message make any sense? Did I understand correctly what is happening and why your changes fix the issue?

Yes. This is what I intended to build.

However, while this indeed fixes the problem, it will lead to oversplitting. Consider the following

```mermaid
graph TD;

subgraph entrypoints
main
end

subgraph fine
app.A
app.B
app.C
end

subgraph coarse
lib.A
lib.B
lib.C
end

main --> app.A
app.A --> app.B --> lib.B --> lib.A
app.A --> app.C --> lib.C --> lib.A
```

With the current algorithm, all of `lib` will be put into a single module.

With the change in this PR, each of `lib`s classes will go into an individual module because they do not have the same tags:

lib.A: main, app.A, app.B, app.C
lib.B: main, app.A, app.B
lib.C: main, app.A, app.C

I'm wondering whether having a single additional tag for the excluded classes is enough (as in: is it downstream of the excluded classes). But I suspect that this is also can lead to cycles. I'll experiment a bit.

File: linker/shared/src/test/scala/org/scalajs/linker/SmallModulesForSplittingTest.scala (hunk 1 above)
Comment (sjrd) [reply linker/shared/src/test/scala/org/scalajs/linker/SmallModulesForSplittingTest.scala:134]: Ah yes, that's simpler indeed. Updated.

**Comment by gzm0** [comment]
Of course, we can break the "single downstream flag" by introducing a hop into our original example:

```mermaid
graph TD;

subgraph entrypoints
main
mod
end

subgraph coarse
lib.B
lib.D
end

subgraph fine
app.A
app.C
end

mod --> app.A
main --> app.A
app.A --> lib.B --> app.C --> lib.D
```

app.C and the coarse module now form a circular dependency :(

**Comment by sjrd** [comment]
Hum, I turned your latest example into a test and it passes:
```scala
    val EntryPointsClass = ClassName("entry.EntryPoints")
    val entryPointsClassDef = classDef(
      EntryPointsClass,
      superClass = Some(ObjectClass),
      methods = List(
        trivialCtor(EntryPointsClass)
      ),
      topLevelExportDefs = List(
        TopLevelMethodExportDef("moda",
            JSMethodDef(SMF, str("expa"), Nil, None, call("app.A", "baz"))(EOH, UNV)),
        TopLevelMethodExportDef("modb",
            JSMethodDef(SMF, str("expb"), Nil, None, call("app.A", "baz"))(EOH, UNV))
      )
    )

    val classDefs = Seq(
      entryPointsClassDef,
      methodHolder("app.A", "baz", BinaryOp(BinaryOp.Int_+, call("lib.B", "bar"), int(2))),
      methodHolder("lib.B", "bar", call("app.C", "foo")),
      methodHolder("app.C", "foo", call("lib.D", "bar")),
      methodHolder("lib.D", "bar", int(1))
    )
```

**Comment by gzm0** [comment]
Yes, it passes with the current solution. But the current solution will oversplit coarse modules, so it is no good.

**Comment by sjrd** [comment]
Ah right, you meant it would fail with:

> I'm wondering whether having a single additional tag for the excluded classes is enough (as in: is it downstream of the excluded classes). But I suspect that this is also can lead to cycles. I'll experiment a bit.

**Comment by sjrd** [comment]
Replaced by #4846.

**Pull request closed without merging**